### DISCUSSION TITLE
Can I use the same position multiple times to display ads in one page?

### DISCUSSION THREAD

**Discussion by giuliopons** [giuliopons/adadmin#37] [Q&A] [answered]
If you need to reuse the same script position in a single page to display ads, be sure to change the id of the ad in the script, because you can’t have the same `<div id="AADIV...">` as target of the script that show banners.

So, when you copy this code remember that in the second, third (etc.) pasted script you have to modify the id name.

The `id` name appears two times "`AADIV34`" in the script, you can see the two strings that you have to change:

`<div id="AADIV34"></div><script>/* Footer prod */ var k=decodeURIComponent(document.cookie),ca=k.split(';'),psc="";for(var i=0;i<ca.length;i++){var c=ca[i];while(c.charAt(0)==' ') c=c.substring(1);if (c.indexOf("adcapban")==0)psc+=(psc==""?"":",")+c.replace("adcapban","").replace("=",",");}var s = document.createElement("script");s.src = "https://www.barattalo.it/amb/ser.php?t=AADIV34"+String.fromCharCode(38)+"f=34"+String.fromCharCode(38)+"psc=" + psc;document.head.appendChild(s);</script>`

**Comment by lamontef** [comment, marked as answer]
What I've found is that this solution only partially works, what has happened with my installation is that when I use this code with the ID changed, I end up with all the same ad on the page. Based on the explanation of how the code works, with the system grabbing ads sequentially from the database, if there are three ads on a page, they all will send a request for the next ad at the same time and will receive the same ad because the pointer to the next ad will not have changed.

I've found that the solution that works for me is to put a delay on each instance of the code (after the first one, that doesn't need a delay). I've had to experiment with the length of the delay for each ad to find the right amount of time to have each delay be able to call the database after the previous ad, yet not delay the loading of the ad too long. I'm still working on it but it seems that 350 milliseconds has been relatively consistent for me.

I still change the ID, so if I have three ads on a page for my track and field website I use TRACK01, TRACK02, and TRACK03.

How I add the delay to the code looks like this:

`<div id="TRACK03"></div><script>/* 300x250 */ setTimeout(function() { var k=decodeURIComponent(document.cookie),ca=k.split(';'),psc="";for(var i=0;i<ca.length;i++){var c=ca[i];while(c.charAt(0)==' ') c=c.substring(1);if (c.indexOf("adcapban")==0)psc+=(psc==""?"":",")+c.replace("adcapban","").replace("=",",");}var s = document.createElement("script");s.src = "https://vcpads.com/amb/ser.php?t=TRACK03"+String.fromCharCode(38)+"f=34"+String.fromCharCode(38)+"psc=" + psc;document.head.appendChild(s);}, 700);</script>`

Since I can't figure out how to format text inside the code block here, I'll have to explain it...

Right after /* 300x250 */ and before var k=decode... I've added - `setTimeout(function() {` 

Right before the </script> close I've added `}, 700);`

The delay length is the number at the end (700 in this case), those are milliseconds.

The first ad on the page doesn't need the delay, the second is delayed 350 milliseconds, and the third 700. I've run a lot of tests on an individual page with various values and am still tweaking it, but this has seemed to work. Since the additional ads are farther down the page, I've yet to see a noticeable delay in the ad display, but I'm sure there's some page load hit. When I started this I found on a page full of ads 50 milliseconds worked well, but when I put it on my actual page, I was still seeing duplicate ads, so I had to keep bumping the delay. I still see some duplicates, but it's a lot better now and I don't want to increase the delay much more. I'm running my pages in WordPress by taking this code into another plug-in that gives me additional flexibility like having a group of ads rotate in an ad space, which causes another challenge for the code (I was getting the same ad five times in a space until I added the delays).

The ultimate solution would be to have the option to select the ads randomly from the database instead of sequentially...

The next challenge is to figure out how to get this code to work on an AMP page, Google flags it as custom javascript and doesn't index the page...

To see the ads in action, here are a couple of my sites (I'm running 18 sites):

https://vcpathletics.com/
https://vcphoops.com/
https://vcpnewz.com/

LaMonte M. Forthun

**Reply by giuliopons** [reply]
Ok, for timeout.

Another way to try to avoid duplicates is to append a random string at the `ser.php` URL in the JavaScript call.
You have duplicates because the browser decides that the page is asking the same resource. 

To avoid duplicates, you also need to have many banners linked to that position, because while your page is loading banners, there could be different concurrent calls to the same position from other users that load a page with same banners. So, their call are mixed.

Here I've also added `Math.random()` on a new `rand` parameter:

`<div id="TRACK03"></div><script>/* 300x250 */ setTimeout(function() { var k=decodeURIComponent(document.cookie),ca=k.split(';'),psc="";for(var i=0;i<ca.length;i++){var c=ca[i];while(c.charAt(0)==' ') c=c.substring(1);if (c.indexOf("adcapban")==0)psc+=(psc==""?"":",")+c.replace("adcapban","").replace("=",",");}var s = document.createElement("script");s.src = "https://vcpads.com/amb/ser.php?t=TRACK03"+String.fromCharCode(38)+"f=34"+String.fromCharCode(38)+"psc=" + psc;document.head.appendChild(s) + String.fromCharCode(38)+"rand=" + Math.random();}, 700);</script>`

I still think that sequential is better than random, why do you think that with random you don't get duplicates?

I've seen the AMP page problem, but can't figure out how to solve it.

**Comment by giuliopons** [comment]
Another trick is this one:

`<div id="AADIV_WILLCHANGE"></div><script>/* 320x250px */ var NewRandomName = 'AADIV_' + Math.floor(Math.random()*10000);document.getElementById("AADIV_WILLCHANGE").setAttribute("id", NewRandomName);var k=decodeURIComponent(document.cookie),ca=k.split(';'),psc="";for(var i=0;i<ca.length;i++){var c=ca[i];while(c.charAt(0)==' ') c=c.substring(1);if (c.indexOf("adcapban")==0)psc+=(psc==""?"":",")+c.replace("adcapban","").replace("=",",");}var s = document.createElement("script");s.src = "https://THISISYOURSITE/amb/ser.php?t="+NewRandomName+String.fromCharCode(38)+"f=9"+String.fromCharCode(38)+"psc=" + psc;document.head.appendChild(s);</script>`

In this code the "Position" used is the number "`9`" (`f=9`) but the target div is dinamically changed. So you can put this code whenever you want in a feed, it automatically changes the target div `AADIV_WILLCHANGE` with a random one.
